### ISSUE TITLE
Improve Keras Policy Memory Usage

### ISSUE THREAD

**Issue by wochinge** [RasaHQ/rasa_core#1451]
Training the demo bot with 15 GiB of Ram failed when the training data was shuffled with a `MemoryError`. 

Concrete function `rasa_core/training/data.py`:
```python
    def shuffled_X_y(self):
        idx = np.arange(self.num_examples())
        np.random.shuffle(idx)
        shuffled_X = self.X[idx]
        shuffled_y = self.y[idx]
        return shuffled_X, shuffled_y
```

Ideas:
* use sklearn.utils.shuffle
* do the shuffling as part of the tensorflow graph

**Comment by akelad** [comment]
@Ghostvv does this make sense to do?

**Comment by Ghostvv** [comment]
I don't think as part of the tf graph makes sense, because we use internal keras to feed bumpy array to tf graph
if sklearn is more efficient then yes

**Comment by akelad** [comment]
I somehow doubt it's going to make a huge difference 🤔

**Comment by wochinge** [comment]
I doubt where sklearn shuffle is more efficient, but worth a try I guess

**Comment by Ghostvv** [comment]
Does keras have internal shuffle? Maybe, we don't need that one

**Comment by wochinge** [comment]
@Ghostvv Yes! Let's do the shuffling as part of the TensorFlow graph, e.g. with https://www.tensorflow.org/api_docs/python/tf/random/shuffle

**Comment by Ghostvv** [comment]
but we don't feed tensors to keras, we feed bumpy arrays. We need to check if keras does it anyhow

**Comment by wochinge** [comment]
we could convert it to tensors (https://www.tensorflow.org/api_docs/python/tf/convert_to_tensor) or have placeholder tensors and then feed the numpy arrays to the placeholders.

**Comment by Ghostvv** [comment]
converting back and forth, I don't think it'll be faster

**Comment by Ghostvv** [comment]
https://stackoverflow.com/questions/50184144/shuffle-in-the-model-fit-of-keras

**Comment by Ghostvv** [comment]
we don't need it at all

**Comment by Ghostvv** [comment]
just set shuffle to true, if it is not set by default
